### ISSUE TITLE
unecessarily slow get_peft_model() with LoRA (esp. in LLMs)

### ISSUE THREAD

**Issue by poedator** [huggingface/peft#871]
### System Info

peft 0.6.0. dev0

### Who can help?

@pacman100 
@BenjaminBossan 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

just measure time it takes to run get_peft_model() with any large LLM and massive peft_config (guanaco style)

### Expected behavior

get_peft_model() with LoRA raking much less time than loading the original model.

**Comment by BenjaminBossan** [comment]
Thank you for reporting this and for digging into the underlying reasons. As you mentioned, there is probably room for improvement here and we'll take a look at your suggestion and what else we can do to speed things up.

**Comment by poedator** [comment]
How this story ended:
the PR https://github.com/huggingface/peft/pull/872 did not go through because a more elegant solution was found.
the solution for the slow init issue was implemented in these PRs: https://github.com/huggingface/peft/pull/887, https://github.com/huggingface/peft/pull/915 by @BenjaminBossan

**Comment by jph00** [comment]
@BenjaminBossan @pacman100 I don't think this issue is resolved yet -- or at least there's a highly related issue. The slowness caused by unnecessary inits also occurs in `merge_and_unload`, via this code path:

```
  File "/usr/local/lib/python3.10/dist-packages/peft/tuners/lora/model.py", line 701, in merge_and_unload
    return self._unload_and_optionally_merge(progressbar=progressbar, safe_merge=safe_merge)
  File "/usr/local/lib/python3.10/dist-packages/peft/tuners/lora/model.py", line 420, in _unload_and_optionally_merge
    new_module = torch.nn.Linear(target.in_features, target.out_features, bias=bias)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/linear.py", line 101, in __init__
    self.reset_parameters()
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/linear.py", line 107, in reset_parameters
    init.kaiming_uniform_(self.weight, a=math.sqrt(5))
```

**Comment by jph00** [comment]
I discovered one possible partial fix to this, which is to add `device=target.weight.device` in `_unload_and_optionally_merge`, in the call `new_module = torch.nn.Linear(target.in_features, target.out_features, bias=bias)`. By putting it on the device, the unnecessary init is at least accelerated.

**Comment by BenjaminBossan** [comment]
@jph00 Indeed there are still parts of the code that have not been optimized for faster init yet. We track the progress in #896. There is already a PR for bnb layers (#994). Also thanks for bringing up `merge_and_unload`, I added it to the list of TODOs.

**Comment by BenjaminBossan** [comment]
Update: The latest PEFT version released today (v0.6.0) includes the speed improvement for bnb LoRA layers.

**Comment by EaswarGn** [comment]
Hi @BenjaminBossan I got this exact same issue in the latest peft version (0.18.1), I opened an issue here: #3039
